csharp/src/Microsoft.ML.OnnxRuntime/ProviderOptions.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntime
+{
+    /// <summary>
+    /// Search Algorithm for Cudnn Conv.
+    /// </summary>
+    public enum OrtCudnnConvAlgoSearch
+    {
+        EXHAUSTIVE,  //!< expensive exhaustive benchmarking using cudnnFindConvolutionForwardAlgorithmEx
+        HEURISTIC,   //!< lightweight heuristic based search using cudnnGetConvolutionForwardAlgorithm_v7
+        DEFAULT,     //!< default algorithm using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM
+    }
+
+    /// <summary>
+    /// Provider options for CUDA EP.
+    /// </summary>
+    public struct OrtCUDAProviderOptions
+    {
+        public int device_id;                                   //!< cuda device with id=0 as default device.
+        public OrtCudnnConvAlgoSearch cudnn_conv_algo_search;   //!< cudnn conv algo search option
+        public UIntPtr gpu_mem_limit;                           //!< default cuda memory limitation to maximum finite value of size_t.
+        public int arena_extend_strategy;                       //!< default area extend strategy to KNextPowerOfTwo.
+        public int do_copy_in_default_stream;                   //!< Whether to do copies in the default stream or use separate streams.
+    }
+
+    /// <summary>
+    /// Holds provider options configuration for creating an InferenceSession.
+    /// </summary>
+    public class ProviderOptions : SafeHandle
```

Comment:
>SafeHandle [](start = 35, length = 10)

You made this class IDisposable and, therefore, it must be disposed, which the codes not do.
Even if it currently does not have a real handle to dispose, in the future, it might have it, but the code will set a bad example of not disposing it.

Reply:
Thanks for reminding the importance of presence of dispose function.
I've added dispose function in this class, even though it currently doesn't have a real handle as you found.

Reply:
I think you need to read on how to implement IDisposable and what SafeHandle is. 
1) What is it that you are disposing? Handle is never initialized to any meaninful value and it is not clear what it represents. Your Dispose() simply sets it to zero and does nothing.
2) SafeHandle already implements Dispose() and it calls ReleaseHandle() so the imlpementor of the class does the custom thing that they need to do. In this case your Dispose() overrides SafeHandle implementation and prevents ReleaseHandle() to be ever called. If it was it would be a never ending recursion.

This class as far as I can see should not inherit from SafeHandle.

---
In reply to: [612546105](https://github.com/microsoft/onnxruntime/pull/7315#discussion_r612546105) [](ancestors = 612546105)

Reply:
The reason I made this class a derived class from SafeHandle is for future use where the class needs to call ort native api and should be released once it's done.
Agree with you that so far this class doesn't need to inherit from SafeHandle. For simplicity, I made this class a simple static class.

Reply:
You should not override Dispose(). If that is removed and ReleaseHandle() continues to call Dispose() it would recurse.

---
In reply to: [613739660](https://github.com/microsoft/onnxruntime/pull/7315#discussion_r613739660) [](ancestors = 613739660)